Support LocalDateTime as search argument in query methods

Is it possible to support LocalDateTime as a search argument in Spring Data query methods?

For example, we have a model:


```
Entity(name = "singers")
public class Singer { 

    @Id     
    @Field(name = "singer_id")
    private String singerId;

    @Field(name = "created_date")
    private LocalDateTime createdDate;
}
```



Datastore repository has a search method:
`List<Singer> findAllByCreatedDateLessThan(LocalDateTime date);`

After calling the method an error occurs:

```
Caused by: org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException: Unable to convert class java.time.LocalDateTime to Datastore supported type.
	at org.springframework.cloud.gcp.data.datastore.core.convert.DatastoreNativeTypes.wrapValue(DatastoreNativeTypes.java:142) ~[spring-cloud-gcp-data-datastore-1.1.0.BUILD-20181129.214831-188.jar:1.1.0.BUILD-SNAPSHOT]
	at org.springframework.cloud.gcp.data.datastore.repository.query.PartTreeDatastoreQuery.lambda$applySelectWithFilter$1(PartTreeDatastoreQuery.java:207) ~[spring-cloud-gcp-data-datastore-1.1.0.BUILD-20181129.214831-188.jar:1.1.0.BUILD-SNAPSHOT]
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[na:na]
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1492) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:550) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260) ~[na:na]
	at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:517) ~[na:na]
	at org.springframework.cloud.gcp.data.datastore.repository.query.PartTreeDatastoreQuery.applySelectWithFilter(PartTreeDatastoreQuery.java:221) ~[spring-cloud-gcp-data-datastore-1.1.0.BUILD-20181129.214831-188.jar:1.1.0.BUILD-SNAPSHOT]
	at org.springframework.cloud.gcp.data.datastore.repository.query.PartTreeDatastoreQuery.applyQueryBody(PartTreeDatastoreQuery.java:160) ~[spring-cloud-gcp-data-datastore-1.1.0.BUILD-20181129.214831-188.jar:1.1.0.BUILD-SNAPSHOT]
	at org.springframework.cloud.gcp.data.datastore.repository.query.PartTreeDatastoreQuery.execute(PartTreeDatastoreQuery.java:129) ~[spring-cloud-gcp-data-datastore-1.1.0.BUILD-20181129.214831-188.jar:1.1.0.BUILD-SNAPSHOT]
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:605) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.lambda$invoke$3(RepositoryFactorySupport.java:595) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:595) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:59) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:93) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.data.repository.core.support.SurroundingTransactionDetectorMethodInterceptor.invoke(SurroundingTransactionDetectorMethodInterceptor.java:61) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at com.sun.proxy.$Proxy67.findAllByCreatedDateLessThan(Unknown Source) ~[na:na]
	at com.example.DatastoreRepositoryExample.lambda$commandLineRunner$0(DatastoreRepositoryExample.java:80) [classes/:na]
	at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:813) [spring-boot-2.1.0.RELEASE.jar:2.1.0.RELEASE]
	... 5 common frames omitted
```

--- a/docs/src/main/asciidoc/datastore.adoc
+++ b/docs/src/main/asciidoc/datastore.adoc
@@ -745,30 +745,8 @@ public class MyApplication {
 
 ----
 
-==== Transactions
-
-These transactions work very similarly to
-those of `DatastoreOperations`, but is specific to the repository's domain type and provides repository functions
-instead of template functions.
-
-For example, this is a read-write transaction:
-
-[source,java]
-----
-
-@Autowired
-DatastoreRepository myRepo;
-
-public String doWorkInsideTransaction() {
-  return myRepo.performTransaction(
-    transactionDatastoreRepo -> {
-      // work with the single-transaction transactionDatastoreRepo here. This is a DatastoreRepository object.
-
-      return "transaction completed";
-    }
-  );
-}
-----
+Repositories allow you to define custom Query Methods (detailed in the following sections) for retrieving, counting, and deleting based on filtering and paging parameters.
+Filtering parameters can be of types supported by your configured custom converters.
 
 ==== Query methods by convention
 
@@ -912,6 +890,31 @@ public interface TraderRepository extends CrudRepository<Trader, String> {
 	// This method uses the query from the properties file instead of one generated based on name.
 	List<Trader> fetchByName(@Param("tag0") String traderName);
 
+}
+----
+
+==== Transactions
+
+These transactions work very similarly to
+those of `DatastoreOperations`, but is specific to the repository's domain type and provides repository functions
+instead of template functions.
+
+For example, this is a read-write transaction:
+
+[source,java]
+----
+
+@Autowired
+DatastoreRepository myRepo;
+
+public String doWorkInsideTransaction() {
+  return myRepo.performTransaction(
+    transactionDatastoreRepo -> {
+      // work with the single-transaction transactionDatastoreRepo here. This is a DatastoreRepository object.
+
+      return "transaction completed";
+    }
+  );
 }
 ----
 

--- a/spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java
+++ b/spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java
@@ -318,7 +318,7 @@ public class DatastoreTemplate implements DatastoreOperations {
 			}
 			Value<?> value;
 			if (persistentProperty.isCollectionLike()) {
-				Iterable<?> iterableVal = (Iterable<?>) ValueUtil.toIterableIfArray(val);
+				Iterable<?> iterableVal = (Iterable<?>) ValueUtil.toListIfArray(val);
 				saveAll(iterableVal);
 				List<KeyValue> keyValues = StreamSupport.stream((iterableVal).spliterator(), false)
 						.map(o -> KeyValue.of(this.getKey(o, false)))
@@ -345,7 +345,7 @@ public class DatastoreTemplate implements DatastoreOperations {
 					if (val != null) {
 						//we can be sure that the property is an array or an iterable,
 						//because we check it in isDescendant
-						saveAll((Iterable<?>) ValueUtil.toIterableIfArray(val), key);
+						saveAll((Iterable<?>) ValueUtil.toListIfArray(val), key);
 					}
 				});
 	}

--- a/spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/DatastoreNativeTypes.java
+++ b/spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/DatastoreNativeTypes.java
@@ -16,10 +16,12 @@
 
 package org.springframework.cloud.gcp.data.datastore.core.convert;
 
+import java.util.Arrays;
 import java.util.Map;
 import java.util.Set;
 import java.util.function.BiFunction;
 import java.util.function.Function;
+import java.util.stream.Stream;
 
 import com.google.cloud.Timestamp;
 import com.google.cloud.datastore.Blob;
@@ -41,6 +43,7 @@ import com.google.cloud.datastore.TimestampValue;
 import com.google.cloud.datastore.Value;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.ImmutableSet;
+import com.google.common.primitives.Booleans;
 
 import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException;
 import org.springframework.data.mapping.model.SimpleTypeHolder;
@@ -99,12 +102,18 @@ public abstract class DatastoreNativeTypes {
 				.put(String[].class,
 						builder -> (s, o) -> builder.setBinding(s, (String[]) o))
 				.put(Long.class, builder -> (s, o) -> builder.setBinding(s, (Long) o))
+				.put(Long[].class, builder -> (s, o) -> builder.setBinding(s, Stream.of((Long[]) o)
+						.mapToLong(Long::longValue).toArray()))
 				.put(long[].class, builder -> (s, o) -> builder.setBinding(s, (long[]) o))
 				.put(Double.class, builder -> (s, o) -> builder.setBinding(s, (Double) o))
+				.put(Double[].class, builder -> (s, o) -> builder.setBinding(s, Stream.of((Double[]) o)
+						.mapToDouble(Double::doubleValue).toArray()))
 				.put(double[].class,
 						builder -> (s, o) -> builder.setBinding(s, (double[]) o))
 				.put(Boolean.class,
 						builder -> (s, o) -> builder.setBinding(s, (Boolean) o))
+				.put(Boolean[].class, builder -> (s, o) -> builder.setBinding(s,
+						Booleans.toArray(Arrays.asList(((Boolean[]) o)))))
 				.put(boolean[].class,
 						builder -> (s, o) -> builder.setBinding(s, (boolean[]) o))
 				.put(Timestamp.class,

--- a/spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/TwoStepsConversions.java
+++ b/spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/TwoStepsConversions.java
@@ -144,7 +144,7 @@ public class TwoStepsConversions implements ReadWriteConversions {
 					"Unexpected property embedded type: " + embeddedType);
 		}
 
-		if ((val instanceof Iterable || val.getClass().isArray())
+		if (ValueUtil.isCollectionLike(val.getClass())
 				&& targetCollectionType != null && targetComponentType != null) {
 			try {
 				List elements = (val.getClass().isArray() ? (Arrays.asList(val))
@@ -263,7 +263,7 @@ public class TwoStepsConversions implements ReadWriteConversions {
 			}
 		}
 
-		val = ValueUtil.toIterableIfArray(val);
+		val = ValueUtil.toListIfArray(val);
 
 		if (val instanceof Iterable) {
 			List<Value<?>> values = new ArrayList<>();
@@ -360,6 +360,9 @@ public class TwoStepsConversions implements ReadWriteConversions {
 
 	@Override
 	public Optional<Class<?>> getDatastoreCompatibleType(Class inputType) {
+		if (DatastoreNativeTypes.DATASTORE_NATIVE_TYPES.contains(inputType)) {
+			return Optional.of(inputType);
+		}
 		return DatastoreNativeTypes.DATASTORE_NATIVE_TYPES.stream()
 				.filter(simpleType ->
 						this.internalConversionService.canConvert(inputType, simpleType)

--- a/spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/util/ValueUtil.java
+++ b/spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/util/ValueUtil.java
@@ -19,14 +19,27 @@ package org.springframework.cloud.gcp.data.datastore.core.util;
 import org.springframework.util.CollectionUtils;
 
 /**
+ * Utility class containing functions for working with types related to Cloud Datastore.
+ *
  * @author Dmitry Solomakha
+ * @author Chengyuan Zhao
+ *
+ * @since 1.1
  */
 public class ValueUtil {
-	public static Object toIterableIfArray(Object val) {
-		if (val != null && val.getClass().isArray() && val.getClass() != byte[].class) {
+	public static Object toListIfArray(Object val) {
+		if (val != null && isArrayOfItems(val.getClass())) {
 			//if a property is an array, convert it to list
 			return CollectionUtils.arrayToList(val);
 		}
 		return val;
+	}
+
+	public static boolean isCollectionLike(Class type) {
+		return Iterable.class.isAssignableFrom(type) || isArrayOfItems(type);
+	}
+
+	private static boolean isArrayOfItems(Class type) {
+		return type.isArray() && type != byte[].class;
 	}
 }

--- a/spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/AbstractDatastoreQuery.java
+++ b/spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/AbstractDatastoreQuery.java
@@ -75,8 +75,8 @@ public abstract class AbstractDatastoreQuery<T> implements RepositoryQuery {
 	 * @param param the raw param
 	 * @return an array of a compatible type.
 	 */
-	protected Object[] convertCollectionParamToCompatibleArray(List param) {
-		List converted = (List) param.stream()
+	protected Object[] convertCollectionParamToCompatibleArray(List<?> param) {
+		List converted = param.stream()
 				.map(x -> this.datastoreTemplate.getDatastoreEntityConverter().getConversions().convertOnWriteSingle(x)
 						.get())
 				.collect(Collectors.toList());

--- a/spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/AbstractDatastoreQuery.java
+++ b/spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/AbstractDatastoreQuery.java
@@ -16,6 +16,7 @@
 
 package org.springframework.cloud.gcp.data.datastore.repository.query;
 
+import java.lang.reflect.Array;
 import java.util.Collections;
 import java.util.List;
 import java.util.stream.Collectors;
@@ -66,6 +67,23 @@ public abstract class AbstractDatastoreQuery<T> implements RepositoryQuery {
 		}
 		return rawResult.stream().map(this::processRawObjectForProjection)
 				.collect(Collectors.toList());
+	}
+
+	/**
+	 * Convert collection-like param from the query method into an array of compatible types
+	 * for Datastore.
+	 * @param param the raw param
+	 * @return an array of a compatible type.
+	 */
+	protected Object[] convertCollectionParamToCompatibleArray(List param) {
+		List converted = (List) param.stream()
+				.map(x -> this.datastoreTemplate.getDatastoreEntityConverter().getConversions().convertOnWriteSingle(x)
+						.get())
+				.collect(Collectors.toList());
+		return converted.toArray(
+				(Object[]) Array.newInstance(converted.isEmpty()
+						? String.class // if there are no items in the param
+						: converted.get(0).getClass(), converted.size()));
 	}
 
 	@VisibleForTesting

--- a/spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/GqlDatastoreQuery.java
+++ b/spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/GqlDatastoreQuery.java
@@ -40,6 +40,7 @@ import org.springframework.cloud.gcp.data.datastore.core.convert.DatastoreNative
 import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException;
 import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreMappingContext;
 import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastorePersistentEntity;
+import org.springframework.cloud.gcp.data.datastore.core.util.ValueUtil;
 import org.springframework.data.repository.query.Parameter;
 import org.springframework.data.repository.query.Parameters;
 import org.springframework.data.repository.query.QueryMethodEvaluationContextProvider;
@@ -221,7 +222,15 @@ public class GqlDatastoreQuery<T> extends AbstractDatastoreQuery<T> {
 		}
 		for (int i = 0; i < tags.size(); i++) {
 			Object val = vals.get(i);
-			DatastoreNativeTypes.bindValueToGqlBuilder(builder, tags.get(i), val);
+			Object boundVal;
+			if (ValueUtil.isCollectionLike(val.getClass())) {
+				boundVal = convertCollectionParamToCompatibleArray((List) ValueUtil.toListIfArray(val));
+			}
+			else {
+				boundVal = this.datastoreTemplate.getDatastoreEntityConverter().getConversions()
+						.convertOnWriteSingle(val).get();
+			}
+			DatastoreNativeTypes.bindValueToGqlBuilder(builder, tags.get(i), boundVal);
 		}
 		return builder.build();
 	}

--- a/spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java
+++ b/spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java
@@ -37,7 +37,7 @@ import com.google.cloud.datastore.StructuredQuery.PropertyFilter;
 
 import org.springframework.cloud.gcp.data.datastore.core.DatastoreQueryOptions;
 import org.springframework.cloud.gcp.data.datastore.core.DatastoreTemplate;
-import org.springframework.cloud.gcp.data.datastore.core.convert.DatastoreNativeTypes;
+import org.springframework.cloud.gcp.data.datastore.core.convert.ReadWriteConversions;
 import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException;
 import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreMappingContext;
 import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastorePersistentEntity;
@@ -181,30 +181,33 @@ public class PartTreeDatastoreQuery<T> extends AbstractDatastoreQuery<T> {
 					.getPersistentProperty(part.getProperty().getSegment()))
 							.getFieldName();
 			try {
+
+				ReadWriteConversions converter = this.datastoreTemplate.getDatastoreEntityConverter().getConversions();
+
 				switch (part.getType()) {
 				case IS_NULL:
 					filter = PropertyFilter.isNull(fieldName);
 					break;
 				case SIMPLE_PROPERTY:
 					filter = PropertyFilter.eq(fieldName,
-							DatastoreNativeTypes.wrapValue(it.next()));
+							converter.convertOnWriteSingle(it.next()));
 					equalityComparedFields.add(fieldName);
 					break;
 				case GREATER_THAN_EQUAL:
 					filter = PropertyFilter.ge(fieldName,
-							DatastoreNativeTypes.wrapValue(it.next()));
+							converter.convertOnWriteSingle(it.next()));
 					break;
 				case GREATER_THAN:
 					filter = PropertyFilter.gt(fieldName,
-							DatastoreNativeTypes.wrapValue(it.next()));
+							converter.convertOnWriteSingle(it.next()));
 					break;
 				case LESS_THAN_EQUAL:
 					filter = PropertyFilter.le(fieldName,
-							DatastoreNativeTypes.wrapValue(it.next()));
+							converter.convertOnWriteSingle(it.next()));
 					break;
 				case LESS_THAN:
 					filter = PropertyFilter.lt(fieldName,
-							DatastoreNativeTypes.wrapValue(it.next()));
+							converter.convertOnWriteSingle(it.next()));
 					break;
 				default:
 					throw new DatastoreDataException(

--- a/spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/repository/query/GqlDatastoreQueryTests.java
+++ b/spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/repository/query/GqlDatastoreQueryTests.java
@@ -22,13 +22,17 @@ import java.util.Optional;
 
 import com.google.cloud.datastore.DoubleValue;
 import com.google.cloud.datastore.GqlQuery;
-import com.google.cloud.datastore.StringValue;
+import com.google.cloud.datastore.LongValue;
 import com.google.cloud.datastore.Value;
 import org.junit.Before;
 import org.junit.Test;
 import org.mockito.Mockito;
 
 import org.springframework.cloud.gcp.data.datastore.core.DatastoreTemplate;
+import org.springframework.cloud.gcp.data.datastore.core.convert.DatastoreCustomConversions;
+import org.springframework.cloud.gcp.data.datastore.core.convert.DatastoreEntityConverter;
+import org.springframework.cloud.gcp.data.datastore.core.convert.ReadWriteConversions;
+import org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions;
 import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreMappingContext;
 import org.springframework.cloud.gcp.data.datastore.core.mapping.Entity;
 import org.springframework.cloud.gcp.data.datastore.core.mapping.Field;
@@ -58,6 +62,10 @@ public class GqlDatastoreQueryTests {
 
 	private DatastoreTemplate datastoreTemplate;
 
+	private DatastoreEntityConverter datastoreEntityConverter;
+
+	private ReadWriteConversions readWriteConversions;
+
 	private DatastoreQueryMethod queryMethod;
 
 	private QueryMethodEvaluationContextProvider evaluationContextProvider;
@@ -66,6 +74,10 @@ public class GqlDatastoreQueryTests {
 	public void initMocks() {
 		this.queryMethod = mock(DatastoreQueryMethod.class);
 		this.datastoreTemplate = mock(DatastoreTemplate.class);
+		this.datastoreEntityConverter = mock(DatastoreEntityConverter.class);
+		this.readWriteConversions = new TwoStepsConversions(new DatastoreCustomConversions(), null);
+		when(this.datastoreTemplate.getDatastoreEntityConverter()).thenReturn(this.datastoreEntityConverter);
+		when(this.datastoreEntityConverter.getConversions()).thenReturn(this.readWriteConversions);
 		this.evaluationContextProvider = mock(QueryMethodEvaluationContextProvider.class);
 	}
 
@@ -94,7 +106,11 @@ public class GqlDatastoreQueryTests {
 				+ "price>@tag6 AND price<=@tag7 )ORDER BY id DESC LIMIT 3";
 
 		Object[] params = new Object[] { "BUY", "abcd",
-				new String[] { "abc123", "abc321" }, new double[] { 8.88, 9.99 }, 3.33,
+				// this is an array param of the non-natively supported type and will need conversion
+				new int[] { 1, 2 },
+				new double[] { 8.88, 9.99 },
+				3, // this parameter is a simple int, which is not a directly supported type and uses
+					// conversions
 				"blahblah", 1.11, 2.22 };
 
 		String[] paramNames = new String[] { "tag0", "tag1", "tag2", "tag3", "tag4",
@@ -134,17 +150,21 @@ public class GqlDatastoreQueryTests {
 
 			assertEquals(params[0], paramMap.get("tag0").get());
 			assertEquals(params[1], paramMap.get("tag1").get());
-			assertEquals(((String[]) params[2])[0],
-					((StringValue) (((List) paramMap.get("tag2").get()).get(0))).get());
-			assertEquals(((String[]) params[2])[1],
-					((StringValue) (((List) paramMap.get("tag2").get()).get(1))).get());
+
+			// custom conversion is expected to have been used in this param
+			assertEquals(1L,
+					(long) ((LongValue) (((List) paramMap.get("tag2").get()).get(0))).get());
+			assertEquals(2L,
+					(long) ((LongValue) (((List) paramMap.get("tag2").get()).get(1))).get());
+
 			assertEquals(((double[]) params[3])[0],
 					((DoubleValue) (((List) paramMap.get("tag3").get()).get(0))).get(),
 					0.00001);
 			assertEquals(((double[]) params[3])[1],
 					((DoubleValue) (((List) paramMap.get("tag3").get()).get(1))).get(),
 					0.00001);
-			assertEquals(params[4], paramMap.get("tag4").get());
+			// 3L is expected even though 3 int was the original param due to custom conversions
+			assertEquals(3L, paramMap.get("tag4").get());
 			assertEquals(params[5], paramMap.get("tag5").get());
 			assertEquals(params[6], paramMap.get("tag6").get());
 			assertEquals(params[7], paramMap.get("tag7").get());

--- a/spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQueryTests.java
+++ b/spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQueryTests.java
@@ -30,8 +30,10 @@ import org.junit.Test;
 import org.mockito.Mockito;
 
 import org.springframework.cloud.gcp.data.datastore.core.DatastoreTemplate;
+import org.springframework.cloud.gcp.data.datastore.core.convert.DatastoreCustomConversions;
 import org.springframework.cloud.gcp.data.datastore.core.convert.DatastoreEntityConverter;
 import org.springframework.cloud.gcp.data.datastore.core.convert.ReadWriteConversions;
+import org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions;
 import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException;
 import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreMappingContext;
 import org.springframework.cloud.gcp.data.datastore.core.mapping.Entity;
@@ -76,7 +78,7 @@ public class PartTreeDatastoreQueryTests {
 		this.spannerTemplate = mock(DatastoreTemplate.class);
 		this.spannerMappingContext = new DatastoreMappingContext();
 		this.datastoreEntityConverter = mock(DatastoreEntityConverter.class);
-		this.readWriteConversions = mock(ReadWriteConversions.class);
+		this.readWriteConversions = new TwoStepsConversions(new DatastoreCustomConversions(), null);
 		when(this.spannerTemplate.getDatastoreEntityConverter())
 				.thenReturn(this.datastoreEntityConverter);
 		when(this.datastoreEntityConverter.getConversions())
@@ -95,7 +97,9 @@ public class PartTreeDatastoreQueryTests {
 						+ "ThanEqualAndIdIsNullOrderByIdDesc");
 		this.partTreeSpannerQuery = createQuery();
 
-		Object[] params = new Object[] { "BUY", "abcd", 8.88, 3.33 };
+		Object[] params = new Object[] { "BUY", "abcd",
+				// this int param requires custom conversion
+				8, 3.33 };
 
 		when(this.spannerTemplate.queryKeysOrEntities(any(), any())).thenAnswer(invocation -> {
 			EntityQuery statement = invocation.getArgument(0);
@@ -103,7 +107,7 @@ public class PartTreeDatastoreQueryTests {
 			EntityQuery expected = StructuredQuery.newEntityQueryBuilder()
 					.setFilter(CompositeFilter.and(PropertyFilter.eq("action", "BUY"),
 							PropertyFilter.eq("ticker", "abcd"),
-							PropertyFilter.lt("price", 8.88),
+							PropertyFilter.lt("price", 8L),
 							PropertyFilter.ge("price", 3.33),
 							PropertyFilter.isNull("id")))
 					.setKind("trades")

